File: src/audio_stegano_desktop/src/AudioStegano.Core/Stego/MessageBlockAutoencoder.cs
using System.Text.Json;

namespace AudioStegano.Core.Stego;

/// <summary>Port of MATLAB <c>feedforwardnet(10)</c> 8-10-8 — <c>train/train_autoencoder.m</c>.</summary>
public sealed class MessageBlockAutoencoder
{
    public const int BlockSize = 8;
    public const int HiddenSize = 10;

    private readonly double[][] _iw;
    private readonly double[][] _lw;
    private readonly double[] _b1;
    private readonly double[] _b2;

    private MessageBlockAutoencoder(double[][] iw, double[][] lw, double[] b1, double[] b2)
    {
        _iw = iw;
        _lw = lw;
        _b1 = b1;
        _b2 = b2;
    }

    public static MessageBlockAutoencoder LoadEmbedded()
    {
        var asm = typeof(MessageBlockAutoencoder).Assembly;
        const string name = "AudioStegano.Core.Stego.trained_autoencoder.json";
        using var stream = asm.GetManifestResourceStream(name)
            ?? throw new InvalidOperationException($"Missing embedded resource {name}");
        using var reader = new StreamReader(stream);
        return FromJson(reader.ReadToEnd());
    }

    public static MessageBlockAutoencoder FromJson(string json)
    {
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        return new MessageBlockAutoencoder(
            ReadMatrix(root, "iw"),
            ReadMatrix(root, "lw"),
            ReadVector(root, "b1"),
            ReadVector(root, "b2"));
    }

    public int[] EncodeRounded(ReadOnlySpan<byte> msgBits)
    {
        if (msgBits.Length % BlockSize != 0)
            throw new ArgumentException($"Length must be a multiple of {BlockSize}");
        var nBlocks = msgBits.Length / BlockSize;
        var outVals = new int[msgBits.Length];
        var x = new double[BlockSize];
        for (var b = 0; b < nBlocks; b++)
        {
            for (var i = 0; i < BlockSize; i++)
                x[i] = msgBits[b * BlockSize + i];
            var y = ForwardColumn(x);
            for (var i = 0; i < BlockSize; i++)
                outVals[b * BlockSize + i] = (int)Math.Round(y[i]);
        }
        return outVals;
    }

    public byte[] DecodeBits(ReadOnlySpan<byte> payload)
    {
        if (payload.Length % BlockSize != 0)
            throw new ArgumentException($"Length must be a multiple of {BlockSize}");
        var nBlocks = payload.Length / BlockSize;
        var outBits = new byte[payload.Length];
        var x = new double[BlockSize];
        for (var b = 0; b < nBlocks; b++)
        {
            for (var i = 0; i < BlockSize; i++)
                x[i] = payload[b * BlockSize + i];
            var y = ForwardColumn(x);
            for (var i = 0; i < BlockSize; i++)
                outBits[b * BlockSize + i] = (byte)((int)Math.Round(y[i]) & 1);
        }
        return outBits;
    }

    public static int XorPayloadBit(int encodedRounded, int keyBit)
    {
        var a = encodedRounded != 0;
        var b = keyBit != 0;
        return a != b ? 1 : 0;
    }

    public static byte[] BuildPayload(int[] encodedRounded, byte[] key)
    {
        if (encodedRounded.Length != key.Length)
            throw new ArgumentException("encoded and key must have the same length");
        var payload = new byte[encodedRounded.Length];
        for (var i = 0; i < encodedRounded.Length; i++)
            payload[i] = (byte)XorPayloadBit(encodedRounded[i], key[i]);
        return payload;
    }

    private double[] ForwardColumn(double[] x)
    {
        var n1 = new double[HiddenSize];
        for (var j = 0; j < HiddenSize; j++)
        {
            var sum = _b1[j];
            for (var i = 0; i < BlockSize; i++)
                sum += _iw[j][i] * x[i];
            n1[j] = Tansig(sum);
        }

        var output = new double[BlockSize];
        for (var k = 0; k < BlockSize; k++)
        {
            var sum = _b2[k];
            for (var j = 0; j < HiddenSize; j++)
                sum += _lw[k][j] * n1[j];
            output[k] = Tansig(sum);
        }
        return output;
    }

    private static double Tansig(double n) =>
        2.0 / (1.0 + Math.Exp(-2.0 * Math.Clamp(n, -500, 500))) - 1.0;

    private static double[][] ReadMatrix(JsonElement root, string name)
    {
        var rows = root.GetProperty(name);
        var matrix = new double[rows.GetArrayLength()][];
        var i = 0;
        foreach (var row in rows.EnumerateArray())
        {
            var cols = row.GetArrayLength();
            matrix[i] = new double[cols];
            var j = 0;
            foreach (var v in row.EnumerateArray())
                matrix[i][j++] = v.GetDouble();
            i++;
        }
        return matrix;
    }

    private static double[] ReadVector(JsonElement root, string name)
    {
        var arr = root.GetProperty(name);
        var vec = new double[arr.GetArrayLength()];
        var i = 0;
        foreach (var v in arr.EnumerateArray())
            vec[i++] = v.GetDouble();
        return vec;
    }
}

File: src/audio_stegano_app/lib/core/stego/message_block_autoencoder.dart
import 'dart:convert';
import 'dart:math' as math;
import 'dart:typed_data';

/// Port of MATLAB `feedforwardnet(10)` 8-10-8 with `tansig` — `train/train_autoencoder.m`.
class MessageBlockAutoencoder {
  final List<List<double>> _iw;
  final List<List<double>> _lw;
  final List<double> _b1;
  final List<double> _b2;

  const MessageBlockAutoencoder._(
    this._iw,
    this._lw,
    this._b1,
    this._b2,
  );

  static const int blockSize = 8;
  static const int hiddenSize = 10;

  factory MessageBlockAutoencoder.fromJson(Map<String, dynamic> json) {
    List<List<double>> readMatrix(String key) {
      final raw = json[key] as List<dynamic>;
      return raw
          .map(
            (row) => (row as List<dynamic>).map((e) => (e as num).toDouble()).toList(),
          )
          .toList();
    }

    List<double> readVector(String key) =>
        (json[key] as List<dynamic>).map((e) => (e as num).toDouble()).toList();

    return MessageBlockAutoencoder._(
      readMatrix('iw'),
      readMatrix('lw'),
      readVector('b1'),
      readVector('b2'),
    );
  }

  factory MessageBlockAutoencoder.fromJsonString(String raw) =>
      MessageBlockAutoencoder.fromJson(
        jsonDecode(raw) as Map<String, dynamic>,
      );

  /// `round(net(double(msg_matrix)))` — column-major 8×N blocks flattened.
  List<int> encodeRounded(Uint8List msgBits) {
    if (msgBits.length % blockSize != 0) {
      throw ArgumentError('msgBits length must be a multiple of $blockSize');
    }
    final nBlocks = msgBits.length ~/ blockSize;
    final out = List<int>.filled(msgBits.length, 0);
    for (var b = 0; b < nBlocks; b++) {
      final x = List<double>.generate(
        blockSize,
        (i) => msgBits[b * blockSize + i].toDouble(),
      );
      final y = _forwardColumn(x);
      for (var i = 0; i < blockSize; i++) {
        out[b * blockSize + i] = y[i].round();
      }
    }
    return out;
  }

  /// `round(net(double(reshape(payload, 8, n_blocks))))` flattened.
  Uint8List decodeBits(Uint8List payload) {
    if (payload.length % blockSize != 0) {
      throw ArgumentError('payload length must be a multiple of $blockSize');
    }
    final nBlocks = payload.length ~/ blockSize;
    final out = Uint8List(payload.length);
    for (var b = 0; b < nBlocks; b++) {
      final x = List<double>.generate(
        blockSize,
        (i) => payload[b * blockSize + i].toDouble(),
      );
      final y = _forwardColumn(x);
      for (var i = 0; i < blockSize; i++) {
        out[b * blockSize + i] = y[i].round().clamp(0, 255) & 1;
      }
    }
    return out;
  }

  List<double> _forwardColumn(List<double> x) {
    final n1 = List<double>.filled(hiddenSize, 0);
    for (var j = 0; j < hiddenSize; j++) {
      var sum = _b1[j];
      for (var i = 0; i < blockSize; i++) {
        sum += _iw[j][i] * x[i];
      }
      n1[j] = _tansig(sum);
    }

    final out = List<double>.filled(blockSize, 0);
    for (var k = 0; k < blockSize; k++) {
      var sum = _b2[k];
      for (var j = 0; j < hiddenSize; j++) {
        sum += _lw[k][j] * n1[j];
      }
      out[k] = _tansig(sum);
    }
    return out;
  }

  static double _tansig(double n) =>
      2.0 / (1.0 + math.exp(-2.0 * n.clamp(-500.0, 500.0))) - 1.0;

  /// MATLAB `xor` on numeric values (non-zero is true), then LSB payload bit.
  static int xorPayloadBit(num encodedRounded, int keyBit) {
    final a = encodedRounded != 0;
    final b = keyBit != 0;
    return (a != b) ? 1 : 0;
  }

  /// Build XOR payload from `round(net(...))` and logistic key (embed path).
  static Uint8List buildPayload(List<int> encodedRounded, Uint8List key) {
    if (encodedRounded.length != key.length) {
      throw ArgumentError('encoded and key must have the same length');
    }
    final payload = Uint8List(encodedRounded.length);
    for (var i = 0; i < encodedRounded.length; i++) {
      payload[i] = xorPayloadBit(encodedRounded[i], key[i]);
    }
    return payload;
  }
}
